src/wast-parser.cc

```diff
@@ -1335,6 +1349,21 @@ Result WastParser::ParseModuleFieldList(Module* module) {
       CHECK_RESULT(Synchronize(IsModuleField));
     }
   }
+
+  Result result = Result::Ok;
+
+  for (auto it : context_.resolve_types) {
```

Comment:
This is the core part. We resolve all names after the module is parsed in a single loop.